### ISSUE TITLE
Integrity check error

### ISSUE THREAD

**Issue by isermann** [open-keychain/open-keychain#1322]
Hey, I assume I get this OpenPGP Error: Integrity check, which blocks decrypting.

I'm using German language in K-9 Mail where it's called: "OpenPGP Fehler: Fehlende Integritätsprüfung Dies kann pas...[text is off-screen]" and I get this error since version 3.2.0-3.2.3.
Version 3.1 was working flawlessly, though. Error occurs when receiving emails in K-9 encrypted with Gpg4Win to just one recipient.
I think it's similar to Issue #1016 Can you please look into it again?

**Comment by dschuermann** [comment]
Which version of GPG4Win is used? Does this happen with only one of your contacts? If yes, what's the key ID?

**Comment by isermann** [comment]
It happens with two contacts (I can only test two) using Gpg4Win GPGOL 2.2.2 - 2.2.4 on Windows 7 Outlook 2010.
Though, I can decrypt emails in K-9 send through Mailvelope 0.13.1 to my very same recipient.

**Comment by dschuermann** [comment]
Can you tell us the key id of these contacts? (Go to key -> click on "extended information" inside the menu -> key id is displayed below Name).

**Comment by isermann** [comment]
Emails from this IDs sending encrypted emails from Outlook results in the integrity check error in K-9 with latest Open-Keychain 3.2.3:
Key ID1: 7db8 90eb 76e8 294c
Key ID2: ecee d438 82ec 9c75

**Comment by Valodim** [comment]
Hmm, those keys look fine. Could you send/forward me any email encrypted to one of these keys to look@my.amazin.horse? I won't be able to decrypt it obviously, but the structure might hint what the problem is.

**Comment by isermann** [comment]
Ok, I send you an email encrypted by Key ID2 for my K-9 email recipient.

**Comment by Valodim** [comment]
Received the mail, the data is encrypted without a modification detection packet. This mode of encryption has been deemed insecure and deprecated for more than ten years at this point, and it is very unlikely that any OpenPGP implementation which is still used today does not actually support it. Because of this status, we decided not to support this mode of encryption anymore. However, gpg still encrypts data in this way if any recipient key does not indicate its support of the feature.

Your first key (7db8 90eb 76e8 294c) does have the flag to indicate it supports modification detection, for the second I couldn't check because the key is not on the keyservers. You can send the public key data to me if you'd like me to check and make sure (look@my.amazin.horse). The problematic key here might be another one though, which is 4fa4 6aa5 c4ed cafb. This one does not have the flag, and was one of the recipients in the data you sent to me.

As a simple fix, you can follow these steps in OpenKeychain to make sure your keys have the flag:

- Edit the key (eceed43882ec9c75) in OpenKeychain (pencil button upper right when viewing the key)
- Click the primary subkey, select "Change Expiry" and set it to the same expiry as before (or "No Expiry")
- Save the key

This does not change the expiry, but still updates the key's self-signature which adds the missing flag and in this way repairs the key. It is not a harmful operation otherwise.

It might be a good idea to detect this sort of problem in a key on our side and indicate it to the user, you are likely not the only one affected by this.

**Comment by isermann** [comment]
Thank you for your detailed reply, Valodim!

I tried your simple fix two times by changing the expiry date on key ecee d438 82ec 9c75 and 4fa4 6aa5 c4ed cafb + saving it. But unfortunately it didn't resolve the problem. I'm going to send you the public key data for ecee d438 82ec 9c75.

Is there any way to get apk version 3.1 of open-keychain without compiling it by myself? I remember encryption was working before I updated to version 3.2.

**Comment by Valodim** [comment]
the other key is fine as well, only 4fa4 6aa5 c4ed cafb lacks the flag. note that the fix I mentioned above will fix encrypting a message to the newly fixed key, so you would have to use the updated the key in enigmail to send a message, and that should work then. older messages will not be fixed.

the thing is that we can technically decrypt the data, but it is not up to today's security standards. this leaves us with three options:
- display the message without a warning. this is what enigmail does, and it is quite simply wrong and dangerous
- display the message, giving a warning. this is what gpg does, granting compatibility at the cost of security (because users ignore warnings). this used to be a legitimate decision for compatibility reasons, but see below.
- treat the message as insecure, because its integrity can not be guaranteed. this is what we do, and it is in my opinion the right decision at this point: an attack has been around since 2002. the openpgp standard states these three steps [page 83](https://tools.ietf.org/html/rfc4880#page-83):
 
1. Implementers treat MDC errors and decompression failures as security problems.
2. Implementers implement Modification Detection with all due speed and encourage its spread.
3. Users migrate to implementations that support Modification  with all due speed.

and this statement was made almost a *decade* ago. backwards compatibility means we never get rid of historical security problems.

**Comment by isermann** [comment]
This very interesting. I remember creating this key almost 2 years ago with Mailvelope. Why are they making vulnerable keys almost 11 years after an attack was possible...

Anyway, I tried sending new emails with the by Open-Keychain changed public key from Outlook. When I follow your instructions: change date, save, export, import to Kleopatra. I'm not able to use this key any more for decryption, Kleopatra doesn't offer me to choose it while sending the mail.
If I then change date in Kleopatra for this key (old version not the exported changed one from Open-KC) integrity check fails again in K-9.

Now I created a new key in Kleopatra which now is working in K-9.

Is there any way to edit the old public key to make it work in Kleopatra, maybe the export function of Open-Keychain is not compatible with Kleopatra?
